src/java/org/apache/cassandra/config/YamlConfigurationLoader.java

```diff
@@ -163,21 +169,63 @@ private static void maybeAddSystemProperties(Object obj)
         if (CassandraRelevantProperties.CONFIG_ALLOW_SYSTEM_PROPERTIES.getBoolean())
         {
             java.util.Properties props = System.getProperties();
-            Map<String, String> map = new HashMap<>();
-            for (String name : props.stringPropertyNames())
+            Map<String, Object> map = new HashMap<>();
+            for (String originalKey : props.stringPropertyNames())
             {
-                if (name.startsWith(SYSTEM_PROPERTY_PREFIX))
+                if (originalKey.startsWith(SYSTEM_PROPERTY_PREFIX))
                 {
-                    String value = props.getProperty(name);
-                    if (value != null)
-                        map.put(name.replace(SYSTEM_PROPERTY_PREFIX, ""), value);
+                    String value = props.getProperty(originalKey);
+                    String configKey = originalKey.replace(SYSTEM_PROPERTY_PREFIX, "");
+                    if (value != null && !map.containsKey(configKey)) {
+                        if (!DatabaseDescriptor.hasLoggedConfig()) // CASSANDRA-9909: Avoid flooding config during initialization
+                            logger.warn("Detected JVM property {}={} override for cassandra configuration '{}' (ignored if setting does not exist).", originalKey, value, configKey);
+                        map.put(configKey, getScalarValueOrJsonObject(value));
+                    }
                 }
             }
             if (!map.isEmpty())
                 updateFromMap(map, false, obj);
         }
     }
 
+    private static void maybeAddEnvironmentVariables(Object obj)
+    {
+        if (CassandraRelevantProperties.CONFIG_ALLOW_ENVIRONMENT_VARIABLES.getBoolean())
+        {
+            Map<String, String> environment = System.getenv(); // checkstyle: suppress nearby 'blockSystemPropertyUsage'
+            Map<String, Object> configOverrides = new HashMap<>();
+            for (Map.Entry<String, String> env : environment.entrySet())
+            {
+                String originalKey = env.getKey();
+                if (env.getKey().startsWith(ENVIRONMENT_VARIABLE_PREFIX))
+                {
+                    String configKey = LocalizeString.toLowerCaseLocalized(originalKey.replace(ENVIRONMENT_VARIABLE_PREFIX, "")
+                                                                                      .replace(NESTED_CONFIG_SEPARATOR_ENVIRONMENT, NESTED_CONFIG_SEPARATOR));
+                    String configValue = env.getValue();
+                    // TODO: do not include config if it is not a valid config key
```

Comment:
I am mixed about this.  The config validation will yell that you did this, but you won't know its coming from YAML vs Prop vs Env....

Reply:
The settings table uses

```
Map<String, Property> properties = loader.flatten(Config.class);
```

this tells you every nested property that is allowed.

Reply:
Updated 6fdfc52ddd to allow only overriding only valid properties (nested and top-level)